src/Compilers/CSharp/Portable/BoundTree/BoundExpression.cs

```diff
@@ -280,11 +280,18 @@ public override Symbol ExpressionSymbol
 
     internal sealed partial class BoundDeconstructionAssignmentOperator : BoundExpression
     {
-        internal class AssignmentInfo
+        internal class AssignmentStep
         {
             public BoundAssignmentOperator Assignment;
-            public BoundLValuePlaceholder LValuePlaceholder;
-            public BoundRValuePlaceholder RValuePlaceholder;
+            public BoundDeconstructValuePlaceholder OutputPlaceholder;
+            public BoundDeconstructValuePlaceholder TargetPlaceholder;
+        }
+
+        internal class DeconstructStep
+        {
+            public MethodSymbol DeconstructMemberOpt; // the deconstruct member, or null if tuple deconstruction
+            public BoundDeconstructValuePlaceholder TargetPlaceholder;
+            public ImmutableArray<BoundDeconstructValuePlaceholder> OutputPlaceholders; // placeholders for the outputs produced by this deconstruction
         }
```

Comment:
Yes, I thought about `struct` too. I have a note on that #11299. I think I'll just make the change.

Also, from our discussion, I will move back from `OneOrMany` to simply use `ImmutableArray`.